### PR TITLE
Compress exported mmcif on-the-fly

### LINKED ISSUE #1480: On-the-fly compression for dials.export of unmerged mmcif
> I guess if xia2 is going to write a compressed mmcif it would be make sense if DIALS writes the file as present, and xia2 does the compression?

Ideally, `dials.export` could compress on-the-fly depending on the file extension - e.g. if you pass `mtz.hklout=foo.mmcif.gz` 
or `mtz.hklout=foo.mmcif.bz2` then it would output a compressed file as appropriate.

_Originally posted by @rjgildea in https://github.com/dials/dials/issues/1457#issuecomment-715055863_

### PR DESCRIPTION
**Pull request by dagewa**
Fix #1480

### CODE REVIEW

**Comment by Anthchirp** [comment]
Not quite happy with the automagic here.
Would it be more helpful to have an explicit `compress=gz`/`compress=bz2` option (which may be set by default)?
At the open stage we can the check that the relevant extension is in the name, and if not then append it.

My main issues are with 
* compatibility: (ie. DIALS 2.2/3.2 allow you to write plain text to a file ending in `.bz2`, no problem), and
* expectation and information (user is not told that writing to a `.bz2` file results in a compressed file, whereas writing to a `.zip` file does not)

**Comment by dagewa** [comment]
Explicit compress sounds okay. I mean, we have to explicitly set the `format=mmcif` anyway, not detect based on filename (which I did find annoying already - see discussion at https://github.com/dials/dials/issues/1457#issuecomment-714320850). However, happy to make that change if there is distaste for automagic.

**Comment by dagewa** [comment]
Maybe just `compress=True` and always make it gzip? @rjgildea points out that the PDB uses .gz for downloads

**Comment by graeme-winter** [comment]
> Maybe just `compress=True` and always make it gzip? @rjgildea points out that the PDB uses .gz for downloads

Since this is never (I suggest) going to be used by the "woman in the street" I would say that we should be explicit in preference to implicit - `compress=gz` as default with `bz2` or whatever as options would be preferable to me, provided we also append `.gz` to the filename unless already specified

**Comment by dagewa** [comment]
Ok, I've done that, but I left the default as no compression. That makes it even more explicit. You get exactly the filename you wanted, unless you want one of the supported compression modes and then you get a standard extension. Like you say, this is generally going to get set within xia2 or dui, not by the human in front of the computer.

File: util/export_mmcif.py [hunk 1]
```diff
@@ -324,10 +324,12 @@ def write(self, experiments, reflections):
...
-        elif filename.endswith(".gz"):
+        if self.params.mmcif.compress is not None:
+            filename += "." + self.params.mmcif.compress
+        if self.params.mmcif.compress == "gz":
             open_fn = gzip.open
+        elif self.params.mmcif.compress == "bz2":
+            open_fn = bz2.open
         else:
             open_fn = open
         with open_fn(filename, "wt") as fh:
```
Comment (Anthchirp) [inline util/export_mmcif.py:335]:
```suggestion
        with open_fn(filename, "w") as fh:
```
`t` just raises questions and is the default anyway

**Review by Anthchirp — APPROVED** [review]

File: util/export_mmcif.py (hunk 1 above)
Comment (dagewa) [reply util/export_mmcif.py:335]: Unfortunately this fails abruptly with the message `a bytes-like object is required, not 'str'`

Comment (dagewa) [reply util/export_mmcif.py:335]: no stacktrace and I've not dug into where this happens, but suspect it is within `iotbx/cif/`

Comment (dagewa) [reply util/export_mmcif.py:335]: Yes, it is here: https://github.com/cctbx/cctbx_project/blob/8737e695766c1c7395da6650e5a89777a2d3f145/iotbx/cif/model.py#L83

Comment (Anthchirp) [reply util/export_mmcif.py:335]: did I ever mention that this print-to-file pattern is so, so, so special...

Leave it then :slightly_smiling_face:

**Pull request merged**